### DISCUSSION TITLE
Wifi & BLE

### DISCUSSION THREAD

**Discussion by btsimonh** [arendst/Tasmota#21526] [Q&A]
I've been doing something evil the last few days.
Building a webcam with bluetooth.

This configuration highlights an issue with ESP32.  Wifi and BLE do NOT play nice.  But worse than 'bad wifi performance', on a 'normal' esp32, I can reliably cause a complete failure of the IP subsystem (no web, no ping until power cycle.  Serial logs OK).
I *thought* that the esp32s3 was better.  It is *better*, but still not good enough (i.e. I have seen the same death of connectivity, just harder to reproduce).

Now, this situation is sad, because I wanted to have temp/hum from an external BLE temp sensor, as well as webcam recording to SD.  I assume even MQTT will eventually hit a conflict point and kill network access.

So, Question:
1/ Is there something we can do to improve 'coexistence'.  Reading the docs, it seems that they don't guarantee anything...
We could live with poor performance, but not network death.
2/ does anyone have ideas about what could cause the network death?  Are we using the right libraries for wifi/webserver?
3/ should we abandon BLE completely on ESP32 tasmota, and always recommend a separate device/radio (could be a variant of TAS connected by serial...)?

**Comment by Jason2866** [comment]
Try latest development. We reverted a [PR in Arduino Core](https://github.com/arendst/Tasmota/pull/21524) which has a bad impact for wifi especially UDP traffic.
BUT esp32 with active WiFi and BLE will never work well. Sharing the needed hardware will always be a problem since during WiFi is active BLE telegrams are missed. To improve the situation a perfect WiFi connection should be achieved. So the time where WiFi is active is short. Furthermore the esp32 should have as less as possible load to switch as fast as possible the tasks. So trying to combine CPU eating tasks (like Webcam) and BLE is a bad idea.

**Reply by btsimonh** [reply]
thanks @Jason2866 
hmm :).  
I recompiled this morning with latest dev merged (I have some small webcamv2 sending changes under way).

I see we're up to esp-idf 5.2.1
On my esp32cam ai-thinker style the wifi still dies when under stress (or rather, randomly.  I had a 10 minute run delivering video to the browser, no issue until I opened another browser window.  But then three runs where it died within a few seconds.)
I guess I'll try again each major idf change we do... and until it achieves stable performance not consider mixing the two.
There is an outstanding IDF issue https://github.com/espressif/esp-idf/issues/13734 which could be similar.

**Comment by Jason2866** [comment]
To be correct to compile Arduino 3.0.0 IDF 5.1.4 is used.

**Reply by btsimonh** [reply]
thankyou, I don't know where I got 5.2.1 from - probably from platform-espressif32 release notes, assuming we'd gone to the latest..  but as the commit was a revert, I guess we're not on the latest there.
It is possible that when arduino gets 5.2.1, some fixes there may help.  But probably not.  I'm also considering that it could also be heat related - as with BLE enabled at the moment, it's hard to navigate to the configuration menu to turn it off before it dies :(.  whereas first thing this morning, on the first run, it was more stable.

**Comment by Jason2866** [comment]
Found bugs in IDF master are usually backported to older IDF versions (which are still in support).
IDF 5.1.x is in support. I do updates of the Tasmota Arduino framework when mandantory bug fixes are done (IDF / Arduino). That's the benefit of using our own framework build.

**Comment by raphael-bmec-co** [comment]
We have got BLE and WiFi working fairly well together. We did so using both cores of the ESP32 and letting the primary core focus on just BLE and WiFi and letting the other core do everything else. Having said that, a video stream over WiFi at the same time may be a big ask.

What is the crash log?

**Reply by btsimonh** [reply]
It does not 'crash' as such - but wifi stops operating.  serial logs continue, but there is no response to ping, for example.
I have not attempted to understand more about how to enable more diagnostics at the lower level in IDF which may highlight more about what was happening.
When I originally put together the 'new' BLE driver (to enable drivers like the EQ3 driver), Tasmota delivered for months at a time without major issues, so in less stressed environments, it's most likely not too bad.
A kind of related issue in idf: https://github.com/espressif/esp-idf/issues/13721

I'm note sure it's something they can solve easily with the shared radio, so just accept that BLE and webcam with wifi delivery should probably not be turned on at the same time - unless you want to deliberately cause this kind of problem to try to diagnose it!

**Reply by raphael-bmec-co** [reply]
Hard to tell if it is related, but we've build a supervisory stack to run over the top of the wifi stack - we found without this, there were consistently issues WiFi disconnecting and never connecting again without a reset. Good luck!
